std/net/punycode.d

```diff
@@ -0,0 +1,232 @@
+		throw new InvalidPunycodeException("Invalid Punycode");
+	}
+
+	dchar[] ret;
+	dchar n = initialN;
+	uint i = 0;
+	uint bias = initialBias;
+	dstring dstr = str.to!dstring;
+	assert(dstr.length <= uint.max);
+	immutable ptrdiff_t delimIdx = dstr.lastIndexOf('-');
+	if (delimIdx != -1)
+	{
+		enforce!InvalidPunycodeException(dstr[0 .. delimIdx].all!isASCII, "Invalid Punycode");
+		ret = dstr[0 .. delimIdx].dup;
+	}
+	ptrdiff_t idx = (delimIdx == -1 || delimIdx == 0) ? 0 : delimIdx + 1;
+	while (idx < dstr.length)
+	{
+		immutable uint oldi = i;
+		uint w = 1;
+		for (uint k = base;;k += base)
+		{
+			enforce!InvalidPunycodeException(idx < dstr.length);
+			immutable digit = decodeDigit(dstr[idx]);
+			idx++;
+			enforce!PunycodeException(digit * w <= uint.max - i, "Arithmetic overflow");
+			i += digit * w;
+			immutable t = k <= bias ? tmin :
+				k >= bias + tmax ? tmax : k - bias;
+			if (digit < t) break;
+			enforce!PunycodeException(w <= uint.max / (base - t), "Arithmetic overflow");
+			w *= base - t;
+		}
+		enforce!PunycodeException(ret.length < uint.max-1, "Arithmetic overflow");
+		bias = adaptBias(i - oldi, cast(uint) ret.length + 1, oldi == 0);
+		enforce!PunycodeException(i / (ret.length + 1) <= uint.max - n, "Arithmetic overflow");
+		n += i / (ret.length + 1);
+		i %= ret.length + 1;
+		(() @trusted => ret.insertInPlace(i, n))();
+		i++;
+	}
+	return ret.to!S;
+}
+
+///
+@safe pure
+unittest
+{
+	assert(punyDecode("maana-pta") == "mañana");
+}
+
+@safe pure
+unittest
+{
+	static void assertConvertible(S)(S plain, S punycode)
+	{
+		assert(punyEncode(plain) == punycode);
+		assert(punyDecode(punycode) == plain);
+	}
+	assertConvertible("", "");
+	assertConvertible("ASCII0123", "ASCII0123-");
+	assertConvertible("Punycodeぴゅにこーど", "Punycode-p73grhua1i6jv5d");
+	assertConvertible("Punycodeぴゅにこーど"w, "Punycode-p73grhua1i6jv5d"w);
+	assertConvertible("Punycodeぴゅにこーど"d, "Punycode-p73grhua1i6jv5d"d);
+	assertConvertible("ぴゅにこーど", "28j1be9azfq9a");
+	assertConvertible("他们为什么不说中文", "ihqwcrb4cv8a8dqg056pqjye");
+	assertConvertible("☃-⌘", "--dqo34k");
+	assertConvertible("-> $1.00 <-", "-> $1.00 <--");
+	assertThrown!InvalidPunycodeException(punyDecode("aaa-*"));
+	assertThrown!InvalidPunycodeException(punyDecode("aaa-p73grhua1i6jv5dd"));
+	assertThrown!InvalidPunycodeException(punyDecode("ü-"));
+	assert(collectExceptionMsg(punyDecode("aaa-99999999")) == "Arithmetic overflow");
+}
+
+///	Exception thrown if there was an internal error when encoding or decoding punycode.
+class PunycodeException : Exception
+{
+	import std.exception : basicExceptionCtors;
+    mixin basicExceptionCtors;
+}
+
+/// Exception thrown if supplied punycode is invalid, and therefore cannot be decoded.
+class InvalidPunycodeException : PunycodeException
+{
+	import std.exception : basicExceptionCtors;
+    mixin basicExceptionCtors;
+}
+
+private uint adaptBias(uint delta, in uint numpoints, in bool firsttime) @safe pure nothrow /+@nogc+/
```

Comment:
I think that nogc is an artifact from refactoring. You're right. It can be GC'd.